File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        /*int first=-1;//two pointer at start and end first start with starting if it finds then index and second starts from  ends if it finds then index 
        int last=-1;
        for(int i=0;i<nums.length;i++){
            if(nums[i]==target){
                first=i;
                break;
            }
        }
        for(int i=nums.length-1;i>=0;i--){
            if(nums[i]==target){
                last=i;
                break;
            }
        }
        return new int[]{first,last};*/
        
        int first = findBound(nums, target, true);
        if (first == -1) return new int[]{-1, -1};
        int last = findBound(nums, target, false);
        return new int[]{first, last};
    }

    private int findBound(int[] nums, int target, boolean findFirst) {
        int lo = 0, hi = nums.length - 1;
        int result = -1;

        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;

            if (nums[mid] == target) {
                result = mid; // found a match, remember it

                if (findFirst) {
                    hi = mid - 1; // keep searching LEFT for an earlier one
                } else {
                    lo = mid + 1; // keep searching RIGHT for a later one
                }
            } else if (nums[mid] < target) {
                lo = mid + 1; // target is further right
            } else {
                hi = mid - 1; // target is further left
            }
        }

        return result;
    }
}
    

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.py
class Solution(object):
    def binary_search(self, nums, target,first_find):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        lo=0
        hi=len(nums)-1
        result=-1
        while lo<=hi:
            mid=(lo+hi)//2
            if nums[mid]==target:#ispe aaye the kya kiya humne hum aage piche move krenge to find the value 
                result=mid
                if first_find:
                    hi=mid-1
                else:
                    lo=mid+1
            elif nums[mid]<target:
                lo=mid+1
            else:
                hi=mid-1
        return result

    def searchRange(self,nums,target):
        left =self.binary_search(nums, target, True)
        if left == -1:
            return [-1, -1]
        right = self.binary_search(nums, target, False)
        return [left, right]   